File: internal/resource/wrapper.js
#!/usr/bin/env node

const vm = require('vm');

// Read stdin (JSON payload with code and input)
let inputData = '';
process.stdin.setEncoding('utf8');

process.stdin.on('data', (chunk) => {
  inputData += chunk;
});

process.stdin.on('end', () => {
  try {
    // Parse payload JSON
    const payload = inputData ? JSON.parse(inputData) : {};
    const code = payload.code || '';
    const inputStr = payload.input || '';

    // Parse input JSON
    const event = inputStr ? JSON.parse(inputStr) : {};
    const input = event;

    // Make event and input available globally
    global.event = event;
    global.input = input;

    // Execute user script wrapped so top-level `return` works
    try {
      const wrapped = `(async function(){\n${code}\n})()`;
      const scriptObj = new vm.Script(wrapped, { filename: 'user-code.js' });
      const context = vm.createContext(global);
      Promise.resolve(scriptObj.runInContext(context)).then((res) => {
        if (typeof res !== 'undefined') {
          console.log(JSON.stringify(res));
        }
      }).catch((err) => {
        console.error('Error:', err && err.message ? err.message : String(err));
        process.exit(1);
      });
    } catch (e) {
      console.error('Error:', e && e.message ? e.message : String(e));
      process.exit(1);
    }
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
});

File: internal/resource/wrapper.py
#!/usr/bin/env python3

import sys
import json

# Read stdin (JSON payload with code and input)
input_data = sys.stdin.read()

try:
    # Parse payload JSON
    payload = json.loads(input_data) if input_data.strip() else {}
    code = payload.get('code', '')
    input_str = payload.get('input', '')
    
    # Parse input JSON
    event = json.loads(input_str) if input_str.strip() else {}
    input_var = event

    # Make event and input available globally
    globals()['event'] = event
    globals()['input'] = input_var

    # Execute user script wrapped in a function so top-level `return` works
    func_code = 'def __user_main__():\n'
    for line in code.splitlines():
        func_code += '    ' + line + '\n'

    exec(func_code, globals())
    result = globals()['__user_main__']()

except Exception as e:
    print(f'Error: {str(e)}', file=sys.stderr)
    sys.exit(1)

# If the script returned a value or set a `result`/`__return__` global, print it as JSON
try:
    if 'result' in globals():
        print(json.dumps(globals()['result']))
    elif '__return__' in globals():
        print(json.dumps(globals()['__return__']))
    else:
        # prefer returned value if available
        try:
            print(json.dumps(result))
        except Exception:
            pass
except Exception:
    # ignore serialization errors; leave any prints as-is
    pass
